### ISSUE TITLE
Instantiation error: Arity mismatch

### ISSUE THREAD

**Issue by dhil** [links-lang/links#132]
After merging with `sessions` branch, the following example causes an `Instantiate.ArityMismatch` error to occur in the `effects+sessions` branch:
```
fun state(m) {
  handle(m) {
    case Return(x) -> fun(s) { x }
    case Get(k)      -> fun(s) { k(s)(s) }
    case Put(p,k)    -> fun(s) { k(())(p) }
  }
}

fun runState(m, s0) { 
  var f = state(m); 
  f(s0)
}
```
See stack trace:
```
links> fun runState(m, s0) { var f = state(m); f(s0) };
Arity mismatch in type application
Expression: `TAppl ((`TAbstr ({contents = [(7041, (`Unl , `Any ), `Type ...);
                               (7042, (`Unl , `Any ), `Row ...);
                               (7043, (`Unl , `Any ), `Type ...)]},
                   (`TAppl ((`Var "state", ...),
                             [`Type `MetaTypeVar ...; `Row ({}, ..., false);
                              `Type `MetaTypeVar ...; `Presence `Var ...;
                              `Presence `Var ...]), ...)), ...),
         [`Type `MetaTypeVar ...; `Row ({}, ..., false);
          `Type `MetaTypeVar ...; `Presence `Var ...; `Presence `Var ...])
Type: (() {Get:a,Put:(a) {}-> ()|b}~> c) {Get{d},Put{e}|b}~> (a) {Get{d},Put{e}|b}~> c
Raised at file "transformSugar.ml", line 121, characters 6-37
Called from file "transformSugar.ml", line 554, characters 24-36
Called from file "transformSugar.ml", line 339, characters 26-34
Called from file "transformSugar.ml", line 554, characters 24-36
Called from file "transformSugar.ml", line 345, characters 26-34
Called from file "transformSugar.ml", line 554, characters 24-36
Called from file "transformSugar.ml", line 350, characters 26-34
Called from file "transformSugar.ml", line 554, characters 24-36
Called from file "transformSugar.ml", line 339, characters 26-34
Called from file "transformSugar.ml", line 554, characters 24-36
Called from file "transformSugar.ml", line 345, characters 26-34
Called from file "transformSugar.ml", line 554, characters 24-36
Called from file "transformSugar.ml", line 315, characters 26-34
Called from file "transformSugar.ml", line 554, characters 24-36
Called from file "transformSugar.ml", line 607, characters 24-32
Called from file "transformSugar.ml", line 705, characters 27-35
Called from file "transformSugar.ml", line 736, characters 21-34
Called from file "transformSugar.ml", line 110, characters 23-28
Called from file "transformSugar.ml", line 184, characters 26-59
Called from file "utility.ml" (inlined), line 16, characters 23-26
Called from file "frontend.ml", line 72, characters 24-92
Called from file "frontend.ml", line 28, characters 34-37
Called from file "utility.ml" (inlined), line 16, characters 23-26
Called from file "frontend.ml", line 67, characters 8-417
Called from file "utility.ml" (inlined), line 16, characters 23-26
Called from file "frontend.ml", line 67, characters 8-497
Called from file "utility.ml" (inlined), line 16, characters 23-26
Called from file "frontend.ml", line 67, characters 8-583
Called from file "utility.ml" (inlined), line 16, characters 23-26
Called from file "frontend.ml", line 67, characters 8-657
Called from file "utility.ml" (inlined), line 16, characters 23-26
Called from file "frontend.ml", line 67, characters 8-733
Called from file "utility.ml" (inlined), line 16, characters 23-26
Called from file "frontend.ml", line 67, characters 8-815
Called from file "utility.ml" (inlined), line 16, characters 23-26
Called from file "frontend.ml", line 67, characters 8-899
Called from file "utility.ml" (inlined), line 16, characters 23-26
Called from file "frontend.ml", line 67, characters 8-977
Called from file "utility.ml" (inlined), line 16, characters 23-26
Called from file "frontend.ml", line 67, characters 6-1023
Called from file "links.ml", line 280, characters 34-87
Called from file "links.ml", line 222, characters 20-31
Called from file "camlinternalLazy.ml", line 27, characters 17-27
Re-raised at file "camlinternalLazy.ml", line 34, characters 4-11
Called from file "errors.ml", line 114, characters 4-16
*** Error: Instantiate.ArityMismatch
```

**Comment by dhil** [comment]
I should add that
```
links> state(comp)(0);
```
runs fine assuming `comp` is some stateful computation with integer state.

**Comment by dhil** [comment]
@slindley do you have handle on fixing this?

**Comment by fehrenbach** [comment]
"handle", hah! ;)

**Comment by dhil** [comment]
The following is the smallest example I have been able to construct that reproduces the problem:
```
links> fun h(m) { 
     handle(m) { 
         case Return(x) -> x 
         case Foo(k) -> k(42) 
     } 
};
h = fun : (() {Foo:Int|a}~> b) {Foo{_}|a}~> b

links> fun foo(m) { var f = h(m); f };
Arity mismatch in type application
Expression: `TAppl ((`TAbstr ({contents = [(6871, (`Unl , `Any ), `Row ...);
                               (6872, (`Unl , `Any ), `Type ...)]},
                   (`TAppl ((`Var "h", ...),
                             [`Row ({}, ..., false); `Type `MetaTypeVar ...;
                              `Presence `Var ...]), ...)), ...),
         [`Row ({}, ..., false); `Type `MetaTypeVar ...; `Presence `Var ...])
Type: (() {Foo:Int|a}~> b) {Foo{_}|a}~> b
```
The problem appears whenever `h` is a non-trivial handler (i.e. it contains more than a `Return` clause).

**Comment by dhil** [comment]
The problem seems to be that an additional type argument is introduced somewhere rendering the type application ill-formed:
```
# Type variables (total 2):
1. (6739, (`Unl , `Any ), `Row ...)
2. (6740, (`Unl , `Any ), `Type ...)

# Type arguments (total 3)
1. `Row ({}, ..., false)
2. `Type `MetaTypeVar ...
3. `Presence `Var ...

Arity mismatch in type application
Expression: `TAppl ((`TAbstr ({contents = [(6739, (`Unl , `Any ), `Row ...);
                               (6740, (`Unl , `Any ), `Type ...)]},
                   (`TAppl ((`Var "h", ...),
                             [`Row ({}, ..., false); `Type `MetaTypeVar ...;
                              `Presence `Var ...]), ...)), ...),
         [`Row ({}, ..., false); `Type `MetaTypeVar ...; `Presence `Var ...])
Type: (() {Foo:Int|a}~> b) {Foo{_}|a}~> b
```
Somehow the third type argument `Presence` sneaks its way into the type.

**Comment by dhil** [comment]
Fixed by PR #133.
